api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java

```diff
@@ -217,17 +219,50 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
 				toAdd.add(newObs);
 			}
 		}
-		
+
+		// remove all obs in toRemove recursively from encounter
+		Set<Obs> seenIt = new LinkedHashSet<>();
+		removeAllWithGroupMembers(encounter, toRemove, seenIt);
+
+		// add all obs in toAdd recursively to encounter
+		seenIt = new LinkedHashSet<>();
+		addAllWithGroupMembers(encounter, toAdd, seenIt);
+
+		return encounter;
+	}
+
+	private void removeAllWithGroupMembers(Encounter encounter, List<Obs> toRemove, Set<Obs> seenIt) {
 		for (Obs o : toRemove) {
+			if (seenIt.contains(o)) {
+				continue;
+			}
+			seenIt.add(o);
+			//the above code is to keep same as Encounter.java addObs()
+			//prevent infinite recursion if an obs is its own group member
 			encounter.removeObs(o);
+			if (CollectionUtils.isNotEmpty(o.getGroupMembers(true))) {
+				removeAllWithGroupMembers(encounter, new ArrayList<>(o.getGroupMembers(true)), seenIt);
```

Comment:
Could we assign the new array list to a local variable with a name documenting what it is?